gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/DagManager.java

```diff
@@ -960,49 +962,76 @@ private void checkQuota(DagNode<JobExecutionPlan> dagNode) throws IOException {
+          int userQuotaIncrement = incrementJobCountAndCheckUserQuota(requesterToJobCount, requester.getName(), dagNode);
+          boolean thisRequesterCheck;
+          thisRequesterCheck = userQuotaIncrement < 0;  // user quota check failed
+          usersQuotaIncrement.put(requester.getName(), Math.max(userQuotaIncrement, 0));
+          requesterCheck = requesterCheck && thisRequesterCheck;
+          if (!thisRequesterCheck) {
+            requesterMessage.append(String.format(
+                "Quota exceeded for requester %s on executor %s : quota=%s, runningJobs=%d%n",
+                requester.getName(), specExecutorUri, getQuotaForUser(requester.getName()),
+                requesterToJobCount.get(DagManagerUtils.getUserQuotaKey(requester.getName(), dagNode))));
           }
         }
       }
 
       // Throw errors for reach quota at the end to avoid inconsistent job counts
-      if (!proxyUserCheck) {
-        throw new IOException("Quota exceeded for proxy user " + proxyUser + " on executor " + specExecutorUri +
-            ": quota=" + getQuotaForUser(proxyUser) + ", runningJobs=" + proxyUserToJobCount.get(DagManagerUtils.getUserQuotaKey(proxyUser, dagNode)));
-      }
+      if (!proxyUserCheck || !requesterCheck) {
+        if (!proxyUserCheck) {
+          decrementQuotaUsage(proxyUserToJobCount, proxyUser);
+        }
+
+        if (!requesterCheck) {
+          decrementQuotaUsageForUsers(usersQuotaIncrement);
+        }
 
-      if (!requesterCheck) {
-        throw new IOException(requesterMessage);
+        throw new IOException(requesterMessage.toString());
       }
     }
 
     /**
      * Increment quota by one for the given map and key.
-     * @return true if quota is not reached for this user or user is whitelisted, false otherwise.
+     * @return -1 if quota is already reached for this user, 0 if quota does not need to be increased, +1 if used quota is increased by 1
      */
-    private boolean incrementMapAndCheckQuota(Map<String, Integer> quotaMap, String user, DagNode<JobExecutionPlan> dagNode) {
+    private int incrementJobCountAndCheckUserQuota(Map<String, Integer> quotaMap, String user, DagNode<JobExecutionPlan> dagNode) {
       String key = DagManagerUtils.getUserQuotaKey(user, dagNode);
-      int jobCount = quotaMap.getOrDefault(key, 0);
 
       // Only increment job count for first attempt, since job is considered running between retries
-      if (dagNode.getValue().getCurrentAttempts() == 1) {
-        jobCount++;
-        quotaMap.put(key, jobCount);
+      if (dagNode.getValue().getCurrentAttempts() != 1) {
+        return 0;
       }
 
-      return jobCount <= getQuotaForUser(user);
+      int jobCount = quotaMap.getOrDefault(key, 0);
+      jobCount++;
+
+      if (jobCount > getQuotaForUser(user)) {
+        return -1;
+      }
+
+      quotaMap.put(key, jobCount);
```

Comment:
the same `ConcurrentHashMap::replace` from decrement should be used here too, on the increment.